xbmc/addons/binary-addons/BinaryAddonManager.cpp

```diff
@@ -0,0 +1,237 @@
+    return addon->second;
+
+  CLog::Log(LOGERROR, "CBinaryAddonManager::%s: Requested addon '%s' unknown as binary", __FUNCTION__, addonId.c_str());
+  return nullptr;
+}
+
+AddonPtr CBinaryAddonManager::GetRunningAddon(const std::string& addonId) const
+{
+  CSingleLock lock(m_critSection);
+
+  BinaryAddonBasePtr base;
+  auto addon = m_installedAddons.find(addonId);
+  if (addon != m_installedAddons.end())
+    base = addon->second;
+
+  if (base)
+  {
+    return base->GetActiveAddon();
+  }
+
+  return nullptr;
+}
+
+bool CBinaryAddonManager::AddAddonBaseEntry(BINARY_ADDON_LIST_ENTRY& entry)
+{
+  BinaryAddonBasePtr base = std::make_shared<CBinaryAddonBase>();
+  if (!base->Create(std::get<0>(entry)))
+  {
+    CLog::Log(LOGERROR, "CBinaryAddonManager::%s: Failed to create base for '%s' and addon not usable", __FUNCTION__, std::get<0>(entry).c_str());
+    return false;
+  }
+  m_installedAddons[base->ID()] = base;
+  if (std::get<1>(entry))
+    m_enabledAddons[base->ID()] = base;
+  return true;
+}
+
+void CBinaryAddonManager::OnEvent(const AddonEvent& event)
+{
+  if (auto enableEvent = dynamic_cast<const AddonEvents::Enabled*>(&event))
+  {
+    EnableEvent(enableEvent->id);
+  }
+  else if (auto disableEvent = dynamic_cast<const AddonEvents::Disabled*>(&event))
+  {
+    DisableEvent(disableEvent->id);
+  }
+  else if (typeid(event) == typeid(AddonEvents::InstalledChanged))
+  {
+    InstalledChangeEvent();
+  }
+}
+
+void CBinaryAddonManager::EnableEvent(const std::string& addonId)
+{
+  CSingleLock lock(m_critSection);
+
+  BinaryAddonBasePtr base;
+  const auto& addon = m_installedAddons.find(addonId);
+  if (addon != m_installedAddons.end())
+    base = addon->second;
+  else
+    return;
+
+  CLog::Log(LOGDEBUG, "CBinaryAddonManager::%s: Enable addon '%s' on binary addon manager", __FUNCTION__, base->ID().c_str());
+  m_enabledAddons[base->ID()] = base;
+
+  /**
+   * @todo add way to inform type addon manager (e.g. for PVR) and parts about changed addons
+   *
+   * Currently only Screensaver and Visualization use the new way and not need informed.
+   */
+}
+
+void CBinaryAddonManager::DisableEvent(const std::string& addonId)
+{
+  CSingleLock lock(m_critSection);
+
+  BinaryAddonBasePtr base;
+  const auto& addon = m_installedAddons.find(addonId);
+  if (addon != m_installedAddons.end())
+    base = addon->second;
+  else
+    return;
+
+  CLog::Log(LOGDEBUG, "CBinaryAddonManager::%s: Disable addon '%s' on binary addon manager", __FUNCTION__, base->ID().c_str());
+  m_enabledAddons.erase(base->ID());
+
+  /**
+   * @todo add way to inform type addon manager (e.g. for PVR) and parts about changed addons
+   *
+   * Currently only Screensaver and Visualization use the new way and not need informed.
+   */
+}
+
+void CBinaryAddonManager::InstalledChangeEvent()
+{
+  CSingleLock lock(m_critSection);
+
+  BINARY_ADDON_LIST binaryAddonList;
+  CAddonMgr::GetInstance().GetInstalledBinaryAddonPaths(binaryAddonList);
```

Comment:
@AlwinEsch Can you (as we discussed) change `GetInstalledBinaryAddonPaths` to in addition to the paths return `CAddon`s directly, then pass them on to`CBinaryAddonBase`? This way, parsing of all the common info doesn't have to be duplicated.